New update_me user endpoint

mayoral opened the pull request:
This PR implements #12726.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +30,67 @@ def me
+        render_jsonp(Carto::Api::UserPresenter.new(@user).to_poro)
+      rescue CartoDB::CentralCommunicationFailure => e
+        CartoDB::Logger.error(exception: e, user: @user, params: params)
+        render_jsonp({ errors: "There was a problem while updating your data. Please, try again and contact us if the problem persists" }, 400)
```
Please take this into account ^

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -102,6 +165,29 @@ def uri_user
+        if params[:user_id] == current_viewer.id
+          @user = current_viewer
+        else
+          if @organization.present?
```
Please take this into account (if this line is not deleted) ^

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -102,6 +165,29 @@ def uri_user
       end
+
+      def load_user_to_be_updated
+        @user = nil
```
There's no need for setting instance variables to `nil`.

juanignaciosl commented on config/routes.rb:
```diff
@@ -561,7 +561,12 @@
+      put 'users/:user_id/account' => 'users#update_account', as: :api_v3_users_update_account
+      put 'users/:user_id/profile' => 'users#update_profile', as: :api_v3_users_update_profile
+
+      put 'organization/:id_or_name/users/:user_id/account' => 'users#update_account', as: :api_v3_org_users_update_account
```
Currently, there's no UI for an organization administrator to edit a user profile or account data. Even more: the organization administrator can change users quota, which is not currently supported by this API. @matallo , what do you think about this? I wouldn't add code that is not used (in this case, support for owners to change account or profile), but I don't know your next plans.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -102,6 +165,29 @@ def uri_user
+
+      def load_user_to_be_updated
+        @user = nil
+        load_organization if params[:id_or_name].present?
```
If you use `load_organization` you must take into account its return value, because it might fail after printing output. This should be `return unless load_organization`, I think (check the code).

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -1,15 +1,17 @@
       include FrontendConfigHelper
 
-      ssl_required :show, :me, :get_authenticated_users
+      ssl_required :show, :me, :update_account, :update_profile, :get_authenticated_users
```
Can't we have a single endpoint for both? Both are similar: take a bunch of attributes and set it if they're present. Having a separation makes sense from a UI point of view, but not from an API point of view. cc @matallo

juanignaciosl commented on config/routes.rb:
```diff
@@ -561,7 +561,12 @@
     scope 'v3/' do
       # Front/back split
-      get 'me' => 'users#me', as: :api_v3_users_me
+      get 'me'                     => 'users#me',             as: :api_v3_users_me
```
Next time don't bother on keeping this kind of alignment. It's a pointless legacy hindrance.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,54 @@ def me
+        attributes = params[:user]
+
+        password_change = (attributes[:new_password].present? || attributes[:confirm_password].present?) &&
+          user.can_change_password?
```
^ honor this, please.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,54 @@ def me
+          user.set_fields(attributes, [:email])
+        end
+
+        update_session_security_token(user) if password_change
```
This line should be before email change, because it's related to password change. Maybe you could extract password changing lines to a single method.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,54 @@ def me
+
+        if password_change
+          user.change_password(
+            attributes[:old_password].presence,
```
`presence` is redundant, you already checked `present?`.

juanignaciosl replied:
^ minor

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,54 @@ def me
+        end
+
+        # This fields are optional
+        user.name = attributes.fetch(:name, nil)
```
I think that this is wrong, isn't it? You say that those fields are optional, but you assign them unconditionally. Thus, you'll overwrite current values. In addition, AFAIK, this is going to be used from two pages, so fields won't be there at all.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,54 @@ def me
+        render_jsonp(Carto::Api::UserPresenter.new(user).to_poro)
+      rescue CartoDB::CentralCommunicationFailure => e
+        CartoDB::Logger.error(exception: e, user: user, params: params)
+        render_jsonp({ errors: "There was a problem while updating your data. Please, try again and contact us if the problem persists" }, 400)
```
Please make the sentence shorter and that's it.

juanignaciosl commented on spec/requests/carto/api/users_controller_spec.rb:
```diff
@@ -30,4 +34,132 @@
+        }
+      end
+
+      it 'updates account data for the given user' do
```
Add some test checking the "optionality" of the values. AFAIK (@matallo, please confirm):
- If a field has a value, attribute must be written.
- If a field is `null` or empty string, attribute must be set to `nil` or empty string.
- If a field is not sent, attribute must be kept to the previous value.

juanignaciosl replied:
^ this would be nice

mayoral replied:
It's covered with the `does not update fields not present in the user hash` and the `sets field to nil if key is present in the hash with a nil value` test cases.

juanignaciosl replied:
Yep. My first comment was previous to 11d74134fe55a08e30e992f30f621a8c4cd69a64 commit, I didn't review the (nice) test suite 👍

juanignaciosl commented on spec/requests/carto/api/users_controller_spec.rb:
```diff
@@ -30,4 +34,132 @@
+      end
+    end
+
+
```
^ fix this.

juanignaciosl commented:
Assigning @mayoral again so he can answer some comments that were overlooked. I'll do a CR anyway.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -102,6 +141,21 @@ def uri_user
+
+      def update_password_if_needed(user, attributes)
+        password_change = (attributes[:new_password].present? || attributes[:confirm_password].present?) &&
+          user.can_change_password?
```
^ again (minor)

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,44 @@ def me
+        end
+
+        if attributes[:avatar_url].present? && valid_avatar_file?(attributes[:avatar_url])
+          user.avatar_url = attributes.fetch(:avatar_url, nil)
```
Minor: redundant `fetch` because of previous `present?` check. In addition, I think that `a_hash.fetch(:something, nil)` is always equivalent to `a_hash[:something]`, shorter.

mayoral replied:
👍

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,44 @@ def me
+        end
+
+        fields_to_be_updated = []
+        fields_to_check = [
```
Minor: this can be made constant.

mayoral replied:
Where do you find better to put that constant? In `UsersController` or in the model? Also, I don't know how to name it, maybe some like `FIELDS_UPDATABLE_IN_ME_ENDPOINT`?

juanignaciosl replied:
As it's a list of fields accepted by this endpoint, this controller is the place. I like `UPDATE_ME_FIELDS`: short and mentions the endpoint.

mayoral replied:
Done in c7823e87024c374cd63b158705b112015cca5f1e.

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,44 @@ def me
+          fields_to_be_updated << field if attributes.has_key?(field)
+        end
+
+        user.set_fields(attributes, fields_to_be_updated) if fields_to_be_updated.present?
```
If checking `if fields_to_be_updated.present?` is interesting to avoid unnecessary setting of values, it might also cover the following four lines so user is not checked for validity, updated in central or saved.

mayoral replied:
The email and the avatar can't be covered by `fields_to_be_updated` since they require additional checks. So I can't move that logic inside this if :(.

juanignaciosl replied:
Indeed, I didn't realize it. 👍

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -28,6 +29,44 @@ def me
+          :disqus_shortname, :available_for_hire
+        ]
+
+        fields_to_check.each do |field|
```
You can replace

```
fields_to_be_updated = []
fields_to_check.each do |field|
  fields_to_be_updated << field if attributes.has_key?(field)
end
```

with

```
fields_to_be_updated = fields_to_check.select { |field| attributes.has_key?(field) }
```

mayoral replied:
Done in c7823e87024c374cd63b158705b112015cca5f1e.

javitonino commented:
**Minors**
- [500 error](https://rollbar.com/carto/CartoDB/items/24616/occurrences/30427842379/) when passing an empty JSON
- Generic 400 error when cannot update to central. This should probably be a 50x (server error) or 422 (data error). `{"errors":"There was a problem while updating your data. Please, try again."}`

**Blocker**
- Any update fails with a central communication error https://rollbar.com/carto/CartoDB/items/24617/

🔴

javitonino commented:
Hold on, it seems this may be systems related (and I may be very very unlucky with the load balancer)

javitonino commented:
I updated my original comment with acceptance results. We have a blocker (avatar update does not work) and a lot of little issues that are up for discussion, mainly things that feel weird for an API user.

Back to dev

mayoral commented:
> - 500 error when passing an empty JSON

Fixed. It now returns 403 if the JSON is empty.

> - Generic 400 error when cannot update to central. This should probably be a 50x (server error) or 422 (data error). `{"errors":"There was a problem while updating your data. Please, try again."}`

I've updated it to return 422.

> - The response includes very different fields than the request. It is hard to see if something has changed.

I've created a new `to_me_poro` in the presenter to expose the fields.

> - Having to pass `old_password` to an APi authenticated to API key feels very weird. Password confirmation is also strange. IMHO: the endpoint should just take a single `password` field, the confirmation thing should be validated by frontend.

Hmm... Although I think it's good to do some checks in the frontend side, I also think it's a good idea to send old and new passwords in the API call.

> - [500 error](https://rollbar.com/carto/CartoDB/items/24618/) when updating `avatar_url`

Fixed.

javitonino commented:
retest this please

juanignaciosl commented on app/controllers/carto/api/user_presenter.rb:
```diff
@@ -40,6 +40,23 @@ def to_poro
         poro
       end
 
+      def to_me_poro
```
As soon as we leak page specifics into a generic presenter, the presenter abstraction is broken, so this can't happen. Then, we have two choices: either add a specific presentation at the controller, or modify an existing presentation adding the new properties.

Adding a specific presentation is not desirable, because we want to abstract presentation to controllers as much as possible.

Can we modify an existing presentation? There are three different presentations:

a) `to_poro` (result): meant to be private. Matches, because our current use case is private. 
b) `to_public_poro`: meant to be public, not an option.
c) `to_eumapi_poro`: IMHO this shouldn't even exist.

I'd go for a. Tradeoffs:
- Existing uses get new fields (straightforwards one, not a problem).
- New use case get fields that are not used. A little performance penalty, but nothing actually harmful.
- In return, we reduce the complexity of an user endpoint, because it returns the same fields than others.

mayoral replied:
Done. Implemented A option (merging fields into `to_poro`).

juanignaciosl commented on app/controllers/carto/api/users_controller.rb:
```diff
@@ -102,6 +141,21 @@ def uri_user
+                          user.can_change_password?
+
+        if password_change
+          user.change_password(
```
Although talking to @javitonino made me think twice about password confirmation and old password need, I'd keep this as it is right now:
1.- JS should be not necessary, so the "confirmation feature" should also be server side.
2.- Same motivation for old password. In addition, asking again for the password is a security measure for users that might keep sessions opened (or those that have their cookies stolen) and their password changed by strangers.

So, this is ok.

javitonino commented:
Acceptance 🍏. All editions I tried worked, and the errors are reasonable. I was unable to edit anything I shouldn't (like quota or username).

Taking a wider look, this changes the returned information for a lot of presenter (everything that involves visualization). I'm not sure about adding more fields to the PermissionUserPresenter. This is increasing the size of the response from dashboard searches by 10%, and it is already quite big. Can we avoid this change and fix the test instead?
